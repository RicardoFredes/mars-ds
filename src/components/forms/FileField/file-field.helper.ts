export const dictionary = {
  placeholder: "Toque para escolher um arquivo",
  placeholderImage:
    "https://cdn.mesalva.com/uploads/image/MjAyMi0wNS0xNiAxOTozNDozMiArMDAwMDQ5NTY5MQ%3D%3D%0A.png",
};
