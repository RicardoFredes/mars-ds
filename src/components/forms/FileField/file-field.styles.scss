.file-field {
  $height: 250px;
  border: 1px dashed #637381;
  border-radius: $border-radius-sm;
  position: relative;
  display: flex;
  height: $height;
  flex-direction: column;
  justify-content: center;
  transition: background-color ease 200ms;
  overflow: hidden;
  &__content {

    &__name {
      position: absolute;
      padding: calc($height / 2) $spacing-lg $spacing-lg;
      background-image: linear-gradient(to bottom, transparent 25%, $color-background-primary);
      right: 0;
      bottom: 0;
      left: 0;
    }
    &__remove-button {
      position: absolute;
      top: $spacing-sm;
      right: $spacing-sm;
      z-index: 1;
      padding: 0 $spacing-md;
    }
    &__image {
      display: block;
      margin: 0 auto;
      &--is-file {
        transition: opacity ease 200ms;
        object-fit: cover;
        object-position: top center;
        width: 100%;
        height: 250px;
      }
    }
  }
  &:not(.file-field--disabled):hover {
    border-color: $color-text-primary;
    background-color: $color-action-hover;
    .file-field__content {
      &__image {
        &--is-file {
          opacity: 0.5;
        }
      }
    }
  }
  input {
    cursor: move;
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    opacity: 0;
  }
  &--disabled {
    border-color: $color-action-disabled;
    .file-field {
      &__content {
        opacity: 0.5;
      }
    }
  }
}
