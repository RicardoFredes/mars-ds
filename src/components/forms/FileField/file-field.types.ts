import type { FileInputProps as CommonFileFieldProps } from "@/components/primitives/FileInput";

type FileFieldProps = CommonFileFieldProps;

export type { FileFieldProps };
