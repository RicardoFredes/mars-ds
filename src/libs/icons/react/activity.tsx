/**
 * WARNING!
 * Do NOT edit this file as it is automatically generated
 * run "yarn generate:icons" to update
 */

  
import type { IconLibProps } from "@/libs/icons/icons.types";

const IconActivity = ({ height = 24, width = 24 }: IconLibProps) => (
<svg
  xmlns="http://www.w3.org/2000/svg"
  width={width}
  height={height}
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  strokeWidth="2"
  strokeLinecap="round"
  strokeLinejoin="round"
>
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12" />
</svg>
);
  
export default IconActivity;