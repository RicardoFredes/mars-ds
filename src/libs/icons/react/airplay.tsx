/**
 * WARNING!
 * Do NOT edit this file as it is automatically generated
 * run "yarn generate:icons" to update
 */

  
import type { IconLibProps } from "@/libs/icons/icons.types";

const IconAirplay = ({ height = 24, width = 24 }: IconLibProps) => (
<svg
  xmlns="http://www.w3.org/2000/svg"
  width={width}
  height={height}
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  strokeWidth="2"
  strokeLinecap="round"
  stroke-linejoin="round"
>
  <path d="M5 17H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-1" />
  <polygon points="12 15 17 21 7 21 12 15" />
</svg>
);
  
export default IconAirplay;