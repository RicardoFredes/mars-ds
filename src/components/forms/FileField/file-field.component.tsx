import type { FileFieldProps } from "./file-field.types";

import classNames from "classnames";
import { useState } from "react";

import Loader from "@/components/basics/Loader";
import ToggleButton from "@/components/basics/ToggleButton";
import FileInput from "@/components/primitives/FileInput";
import Caption from "@/components/typographies/Caption";
import Heading, { HeadingSizes } from "@/components/typographies/Heading";
import { cutText } from "@/services/string";

import { dictionary } from "./file-field.helper";

const FileField = ({
  className,
  onSelectFile,
  onUploadFile,
  placeholder = dictionary.placeholder,
  placeholderImage = dictionary.placeholderImage,
  disabled = false,
  ...props
}: FileFieldProps) => {
  const [file, setFile] = useState<File | null>(null);
  const [loading, setLoading] = useState(false);

  const cn = classNames("file-field", className, { "file-field--disabled": disabled });

  const handleSelectFile = (file: File | null) => {
    setFile(file);
    onSelectFile?.(file);
  };

  const handleUploadFile = (file: string | null) => {
    if (!file) return;
    onUploadFile?.(file);
  };

  const handleSelectFileStart = () => setLoading(true);
  const handleSelectFileEnd = () => setLoading(false);

  const removeFile = () => {
    setFile(null);
    onSelectFile?.(null);
    onUploadFile?.(null);
  };

  if (loading) {
    return (
      <div className={cn}>
        <div className="file-field__content file-field__content--is-loading">
          <Loader size="lg" className="color-primary" />
        </div>
      </div>
    );
  }

  return (
    <div className={cn}>
      {file ? (
        <div className="file-field__content">
          <img
            className="file-field__content__image file-field__content__image--is-file"
            src={URL.createObjectURL?.(file)}
          />
          <Caption className="file-field__content__name">{cutText(file.name)}</Caption>
          <ToggleButton
            data-testid="remove-file-button"
            className="file-field__content__remove-button"
            iconName="close"
            variant="text"
            onClick={removeFile}
          />
        </div>
      ) : (
        <div className="file-field__content text-center">
          <img
            className="file-field__content__image file-field__content__image--is-default"
            height={164}
            src={placeholderImage}
          />
          <Heading level={5} size={HeadingSizes.XSmall}>
            {placeholder}
          </Heading>
        </div>
      )}

      <FileInput
        disabled={disabled}
        onSelectFile={handleSelectFile}
        onUploadFile={handleUploadFile}
        onSelectFileStart={handleSelectFileStart}
        onSelectFileEnd={handleSelectFileEnd}
        {...props}
      />
    </div>
  );
};

export default FileField;
