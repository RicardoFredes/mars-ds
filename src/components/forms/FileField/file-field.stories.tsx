import type { ComponentMeta, ComponentStory } from "@storybook/react";

import { action } from "@storybook/addon-actions";

import { dictionary, FIVE_MB } from "@/components/primitives/FileInput";

import FileField from "./file-field.component";

export default {
  title: "Forms/FileField",
  component: FileField,
  argTypes: {
    name: {
      description: "The name of the file field",
      control: { type: "text" },
      defaultValue: undefined,
    },
    maxSize: {
      description: "The maximum size of the file",
      control: { type: "number" },
      defaultValue: FIVE_MB,
    },
    extensions: {
      description: "The extensions allowed for the file",
      control: { type: "array" },
      defaultValue: [],
    },
    placeholder: {
      description: "The placeholder of the file field",
      control: { type: "text" },
      defaultValue: dictionary.placeholder,
    },
    placeholderImage: {
      description: "The placeholder image of the file field",
      control: { type: "text" },
      defaultValue: dictionary.placeholderImage,
    },
    disabled: {
      description: "Whether the file field is disabled",
      control: { type: "boolean" },
      defaultValue: false,
    },
    notifications: {
      description: "The notifications of the file field",
      control: { type: "object" },
      defaultValue: dictionary.notifications,
    },
    onSelectFile: {
      description: "The callback when the file is selected",
      control: { type: "function" },
      action: action("onSelectFile"),
    },
    onUploadFile: {
      description: "The callback when the file is uploaded",
      control: { type: "function" },
      action: action("onUploadFile"),
    },
  },
} as ComponentMeta<typeof FileField>;

const Template: ComponentStory<typeof FileField> = (args) => <FileField {...args} />;

const DefaultArgs = {
  name: "file-name",
  maxSize: FIVE_MB,
  extensions: ["jpg", "jpeg", "png", "gif"],
  placeholder: dictionary.placeholder,
  placeholderImage: dictionary.placeholderImage,
  notifications: dictionary.notifications,
  disabled: false,
};

export const Default = Template.bind({});
Default.args = DefaultArgs;

export const Disabled = Template.bind({});
Disabled.args = { ...DefaultArgs, disabled: true };
