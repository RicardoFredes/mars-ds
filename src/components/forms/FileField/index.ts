/**
 * WARNING!
 * Do not change this file,
 * because it is automatically generated
 */

import FileField from "./file-field.component";
export default FileField;
export * from "./file-field.types";
