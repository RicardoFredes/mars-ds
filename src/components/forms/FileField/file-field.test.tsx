import type { FileFieldProps } from "./file-field.types";

import { render } from "@testing-library/react";

import FileField from "./file-field.component";

const makeSut = (props?: FileFieldProps) => render(<FileField {...props} />);

describe("<FileField>", () => {
  beforeEach(() => jest.clearAllMocks());

  it("should render component with default props", () => {
    const wrapper = makeSut();
    expect(wrapper.container).toMatchSnapshot();
  });
});
