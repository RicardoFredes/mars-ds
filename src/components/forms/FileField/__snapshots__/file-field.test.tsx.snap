// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileField> should render component with default props 1`] = `
<div>
  <div
    class="file-field"
  >
    <div
      class="file-field__content text-center"
    >
      <img
        class="file-field__content__image file-field__content__image--is-default"
        height="164"
        src="https://cdn.mesalva.com/uploads/image/MjAyMi0wNS0xNiAxOTozNDozMiArMDAwMDQ5NTY5MQ%3D%3D%0A.png"
      />
      <h5
        class="heading heading--size-xs"
      >
        Toque para escolher um arquivo
      </h5>
    </div>
    <input
      class="file-input"
      data-testid="file-input"
      type="file"
    />
  </div>
</div>
`;
